Internal: [NA] Tests task - Create E2E tests [ED-11088]

Barcov opened the pull request.

Barcov committed:
Duplicate Repeater, Fix First open Breaking tests,Fix Style tests,

Barcov committed:
Regenerate Screenshots

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
SVG Icon, No Icon tests, Soft Expects

Barcov committed:
Title Text Screenshot of Title only

Barcov committed:
CI Screenshots

Barcov committed:
Content Style Tests - max items expanded, SVG Icons - Enable Advanced Uploads

Barcov committed:
CI Screenshots

Barcov committed:
CI Screenshots

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Content Style, solve detached frame screenshot issue.

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
SVG Icon test

Barcov committed:
SVG Icon test

Barcov committed:
isScreenShotable() to solve flaky waits. Refactor IconTitle Alignment

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
CI Screenshots

Barcov committed:
Solve Style Tab deselection

Barcov committed:
isScreenShottable & uploadSVG to editor-page

Barcov committed:
unskip Header style tests

Barcov committed:
Content Style Tests wait for animation

Barcov committed:
SVG Icon await render

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Tweak uploadSVG

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Tweak uploadSVG

Barcov committed:
Tweak SVG test

Barcov committed:
CI Screenshots

TzviRabinovitch commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:
```diff
@@ -149,48 +149,64 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
+					await expect.soft( item ).not.toHaveAttribute( 'open', '' );
 				}
 
 				// Check manual select of first expand -> first item is open
```
maybe use test.step instead of comments?

TzviRabinovitch commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:
```diff
@@ -149,48 +149,64 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
-				await expect( firstItemID ).not.toEqual( secondItemID );
-				await expect( secondItemID ).not.toEqual( thirdItemID );
-				await expect( thirdItemID ).not.toEqual( firstItemID );
+				// Reset max_items_expanded to default state
```
not sure if the comment is needed here

TzviRabinovitch commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:
```diff
@@ -253,16 +269,76 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
+				nestedAccordionWidgetId = '48f02ad';
+
+			await wpAdmin.enableAdvancedUploads();
+			const editor = await wpAdmin.useElementorCleanPost();
```
useElementorCleanPost might prevent parallel execution ( not used yet )

Barcov replied:
Good idea @TzviRabinovitch, I've removed all usage of this function in NA tests.

louiswol94 commented on tests/playwright/pages/editor-page.js:
```diff
@@ -778,4 +794,34 @@ module.exports = class EditorPage extends BasePage {
+	* Can be used to check for completed rendering. Useful to wait out animations before screenshots and expects.
+	* Should be less flaky than waitForLoadState( 'load' ) in editor where Ajax re-rendering is triggered.
+	*/
+	async isScreenShottable( locator, retries = 3, timeout = 500 ) {
```
@Barcov Nice!

TzviRabinovitch committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Suggested Changes

Barcov committed:
CI Screenshot

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Suggested Changes

Barcov committed:
CI Screenshots

Barcov committed:
CI Screenshots

hein-obox commented:
@Barcov 
Would it be possible to update `loadJsonPageTemplate` and remove the code duplication from `loadTemplate`  and to call the `loadTemplate` inside the `loadJsonPageTemplate` function?

Barcov committed:
Update tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov committed:
Update tests/playwright/pages/editor-page.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov committed:
Update tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov committed:
Update tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov commented on tests/playwright/pages/editor-page.js:
```diff
@@ -778,4 +794,34 @@ module.exports = class EditorPage extends BasePage {
+	* Can be used to check for completed rendering. Useful to wait out animations before screenshots and expects.
+	* Should be less flaky than waitForLoadState( 'load' ) in editor where Ajax re-rendering is triggered.
+	*/
+	async isScreenShottable( locator, retries = 3, timeout = 500 ) {
```
@davseve I like the proposal for broader naming. I've noticed that I started using this to check stability before other expects than Screenshots. Extracting this would be useful for FE as well.

hein-obox replied:
@Barcov Please send me a PM about this, because I have copied and pasted your function into the pro version as well :).

hein-obox replied:
Btw It would be so much better if we would have one testing codebase.

Barcov replied:
@hein-obox & @davseve We'll need to consider this refactor in a new task next sprint.

Barcov committed:
Update tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:
```diff
@@ -355,146 +436,145 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
 			} );
 		} );
 
 		test( 'Nested Accordion Title Icon and Text Vertical Alignment', async ( { browser }, testInfo ) => {
```
@davseve 3 in this case refers to the number of retries.

```
async isScreenShottable( locator, retries = 3, timeout = 500 ) {
```

Barcov commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion.test.js:
```diff
@@ -253,16 +272,77 @@ test.describe( 'Nested Accordion @nested-accordion', () => {
+			await editor.loadJsonPageTemplate( __dirname, 'nested-accordion-title-and-icons', '.elementor-widget-n-accordion' );
+			await editor.closeNavigatorIfOpen();
+
+			await test.step( 'Check that an SVG title icon is displayed', async () => {
```
@davseve We can do that. Drop me a PM and let's discuss the details.

Barcov replied:
@davseve On testing using SVG icons depends on inline_font experiment being enabled.

Barcov commented:
> @Barcov Would it be possible to update `loadJsonPageTemplate` and remove the code duplication from `loadTemplate` and to call the `loadTemplate` inside the `loadJsonPageTemplate` function?

@hein-obox I think we'll have to push this refactoring out to another task to compare pro and core branches and reconcile differences in a next sprint.

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Rename isScreenShottable to isUiStable

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Update Item Title without re-rendering

arielk committed:
Tweak: Increase the resolution of AI upscale/resize option up to 2048px [ED-11342] (#22987)

<!--
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not
include enough information to be reviewed in a timely manner may be
closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:
https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no
spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

This PR can be summarized in the following changelog entry:

*

An explanation of what is done in this PR

*

This PR can be tested by following these steps:

*

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

TzviRabinovitch committed:
Internal: Re-add soundCloud test [ED-11341] (#22990)

Barcov committed:
Extract Helpers

Barcov committed:
Deprecate useElementorCleanPost() to allow Parallel test execution in future

Barcov committed:
Split Test

Barcov committed:
Merge remote-tracking branch 'upstream/main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge remote-tracking branch 'upstream/main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

TzviRabinovitch committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Partial Rerender Title

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

TzviRabinovitch committed:
Changed experiment dependency

TzviRabinovitch committed:
Changed experiment dependency

Barcov committed:
Merge branch 'internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' of github.com:Barcov/elementor into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge branch 'elementor:main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Changed experiment dependency

Barcov committed:
Remove .run

Barcov committed:
Linting Changes

Barcov committed:
Linting Changes

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

# Conflicts:
#	modules/nested-accordion/widgets/nested-accordion.php

Barcov committed:
Linting Changes

TzviRabinovitch committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Test Waiting to avoid Exception on loadJsonPageTemplate

Barcov committed:
Update CI Screenshots

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Wait for PageLoad

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Wait for PageLoad

Barcov committed:
Linting Fixes

Barcov committed:
Fix Container not found

Barcov committed:
Fix Container not found

Barcov committed:
Widget Select instead of inner container.

Barcov committed:
Add EditCurrentPage()

Barcov committed:
Replace gototPostId (Parallel tests) & containers/container error.

Barcov committed:
Replace gototPostId (Parallel tests)

Barcov committed:
Set Default Permalinks

louiswol94 commented on tests/playwright/sanity/modules/nested-accordion/helper.js:
```diff
@@ -0,0 +1,109 @@
+async function expectScreenshotToMatchLocator( fileName, locator ) {
+	expect.soft( await locator.screenshot( {
+		type: 'png',
+	} ) ).toMatchSnapshot( fileName );
```
@Barcov Is this test flaky? It failed once.

Barcov replied:
@louiswol94 This is essentially shorthand for a normal screenshot with png and soft expect dialled in. So could only be flaky if the locator is flaky. 

We'll possibly need to increase the retires for isUIStable() on the failing screenshot if we see this again.

Barcov commented on .gitignore:
```diff
@@ -23,6 +23,7 @@ yarn.lock
 .project/
 .vscode/
 .idea/
+/.run/
```
PhpStorm settings for local runners like grunt
![image](https://github.com/elementor/elementor/assets/130742176/ba9d53d8-3865-4ce3-8e6a-7f61e96ef6cf)

Barcov commented on .run/watch-scripts.run.xml:
```diff
@@ -1,11 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
@louiswol94 This is a Dev config for local runners like grunt in PhpStorm 
![Uploading image.png…]()

louiswol94 replied:
@Barcov Please revert this change. Should not be in this PR :-)

Barcov commented on .run/watch-styles.run.xml:
```diff
@@ -1,11 +0,0 @@
-<component name="ProjectRunConfigurationManager">
```
@louiswol94 As above

louiswol94 replied:
@Barcov Please revert this change. Should not be in this PR :-)

Barcov commented on tests/playwright/pages/editor-page.js:
```diff
@@ -544,6 +560,34 @@ module.exports = class EditorPage extends BasePage {
+		}
+	}
+
+	async getPageID() {
```
@louiswol94 Yes David and I where discussing this and I have a change like that to merge in on a seperate PR.

louiswol94 replied:
@Barcov I think you should just update it in this function already.

Barcov commented on .gitignore:
```diff
@@ -34,3 +35,4 @@ coverage/
 test-results/
 **/storageState.json
 tests/playwright/config/local.config.js
+
```
@louiswol94 Storm change. Reverting...

Barcov committed:
Suggested Changes

Barcov committed:
Suggested Changes

hein-obox commented on modules/nested-accordion/widgets/nested-accordion.php:
```diff
@@ -767,20 +766,34 @@ protected function content_template() {
+				'data-binding-setting': ['item_title'],
+				'data-binding-index': itemCount,
+			});
+
```
@Barcov Shouldn't this be inside another PR?

louiswol94 replied:
@Barcov Please revert this change. Should not be in this PR :-)

louiswol94 commented on modules/nested-accordion/widgets/nested-accordion.php:
```diff
@@ -271,7 +271,6 @@ protected function register_controls() {
 				],
 				'default' => 'div',
 				'separator' => 'before',
-				'render_type' => 'template',
```
@hein-obox No, the fix for that would be the data-bindings on the title text. Barco removed this key on the title tag which is incorrect.

louiswol94 commented on modules/nested-accordion/widgets/nested-accordion.php:
```diff
@@ -808,7 +821,7 @@ protected function content_template() {
 				<summary {{{ view.getRenderAttributeString( itemTitleKey ) }}}>
 					<span class="e-n-accordion-item-title-header">
-						<{{{ titleHTMLTag }}} class="e-n-accordion-item-title-text">
+						<{{{ titleHTMLTag }}} {{{ view.getRenderAttributeString( itemTitleText ) }}}>
```
@Barcov Please revert this change. Should not be in this PR :-)

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Suggested Changes

Barcov committed:
Suggested Changes

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Suggested Changes

Barcov committed:
Suggested Changes

Barcov committed:
CI Screenshots

hein-obox commented on tests/playwright/pages/wp-admin-page.js:
```diff
@@ -166,4 +171,22 @@ module.exports = class WpAdminPage extends BasePage {
 		await this.page.getByRole( 'link', { name: ' Edit with Elementor' } ).click();
 	}
+
+	async setPermalinks( structure = 'plain' ) {
```
Nice!

louiswol94 replied:
@Barcov This function is not used anywhere. Do we need it?

Barcov replied:
@louiswol94 I used it initially in this PR. We'll be using this in Taxonomy filter soon. It's also useful to tests that Hein has written in Mega menu. So I've left it in after refactoring I can remove it and add it in a separate PR at some point if you prefer...

Barcov replied:
@louiswol94 removing form this PR. Refactoring has orphaned it. I'll add it in again on a later PR.

hein-obox commented on tests/playwright/sanity/modules/nested-accordion/helper.js:
```diff
@@ -0,0 +1,109 @@
+async function expectScreenshotToMatchLocator( fileName, locator ) {
+	expect.soft( await locator.screenshot( {
+		type: 'png',
+	} ) ).toMatchSnapshot( fileName );
```
What is causing the problem?

Barcov replied:
@hein-obox  To fix the issue with "exception require container containers" I had to abandon using selectElement() for this particular test. Dymitro suggested we replace this because it's flaking out and causing those console exceptions "container containers required".

Clicking on the widget by ID resulted in the inner container getting selected which makes the test fail.
Workaround was to click on the item-title. But this closes the first Item. So the test failed once after pushing up this change. 

I'm not expecting this to be flaky. The failure here is because of different height like the diff screenshot I added below.

![image](https://github.com/elementor/elementor/assets/130742176/8cb4fda7-a97b-499d-ae17-851e97a3a083)

hein-obox commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js:
```diff
@@ -0,0 +1,230 @@
+				await nestedAccordionItemTitle.first().locator( '.elementor-editor-container-settings' ).click();
+			} );
+
+			await test.step( 'override background and border', async () => {
```
I would expect this to start with a capital letter (Here and everywhere).

hein-obox commented on tests/playwright/sanity/modules/nested-accordion/nested-accordion-style.test.js:
```diff
@@ -0,0 +1,230 @@
+			} );
+		} );
+	} );
+	test( 'Header style tests', async ( { page }, testInfo ) => {
```
I would expect an empty line above the `test()`. Here and everywhere.

louiswol94 commented on tests/playwright/pages/editor-page.js:
```diff
@@ -544,6 +559,24 @@ module.exports = class EditorPage extends BasePage {
+	}
+
+	async getPageIdFromFrontEnd() {
+		const pageId = await this.page.evaluate( () => elementorFrontendConfig.post.id );
```
@Barcov Should `pageId` not be returned?

Barcov replied:
@louiswol94 Nice save, I've combined getPageIdFromFrontEnd() with getPageId() and included the return

davseve commented on tests/playwright/pages/editor-page.js:
```diff
@@ -544,6 +559,24 @@ module.exports = class EditorPage extends BasePage {
+		await this.gotoPostId( postId );
+	}
+
+	async getPageId() {
+		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
+		return pageId;
+	}
+
+	async getPageIdFromFrontEnd() {
+		const pageId = await this.page.evaluate( () => elementorFrontendConfig.post.id );
```
I think we can unite these two functions into one,
 `elementorFrontendConfig.post.id` is also available in the editor. 
```suggestion
	async getPageId() {
		const pageId = await this.page.evaluate( () => elementorFrontendConfig.post.id );
		return pageId;
	}
		
```

Barcov committed:
Update tests/playwright/pages/editor-page.js

Co-authored-by: Davids Seveloff <94905340+davseve@users.noreply.github.com>

Barcov committed:
Suggested Changes

Barcov committed:
Suggested Changes

Barcov committed:
Revert Suggested Change

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

louiswol94 commented on tests/playwright/pages/editor-page.js:
```diff
@@ -544,6 +559,24 @@ module.exports = class EditorPage extends BasePage {
+		await this.gotoPostId( postId );
+	}
+
+	async getPageId() {
+		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
+		return pageId;
+	}
```
```suggestion
	async getPageId() {
		const pageId = await this.page.evaluate( () => elementor.config.initial_document.id );
		return pageId;
	}

```

louiswol94 committed:
Update tests/playwright/pages/editor-page.js

louiswol94 commented on tests/playwright/sanity/modules/nested-accordion/helper.js:
```diff
@@ -0,0 +1,109 @@
+
+	const controlBreakpoint = ( breakpoint.toLowerCase() !== 'desktop' ) ? '_' + breakpoint : '';
+	const locator = await getChoicesButtonSelector( 'item_title_icon_position' + controlBreakpoint, icon[ direction ] );
+	await editor.page.locator( locator ).click();
+}
+async function setTitleHorizontalAlignment( direction, editor, breakpoint = 'desktop' ) {
+	const icon = Object.freeze( {
```
```suggestion
	await editor.page.locator( locator ).click();
}

async function setTitleHorizontalAlignment( direction, editor, breakpoint = 'desktop' ) {
	const icon = Object.freeze( {
```

louiswol94 committed:
Update tests/playwright/sanity/modules/nested-accordion/helper.js

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

louiswol94 commented on tests/playwright/sanity/modules/nested-accordion/helper.js:
```diff
@@ -0,0 +1,110 @@
+	async function setState() {
+		await editor.page.click( '.elementor-control-accordion_' + state + '_border_and_background' );
+	}
+}
+async function setIconColor( editor, state, color, context ) {
```
```suggestion
}

async function setIconColor( editor, state, color, context ) {
```

louiswol94 committed:
Update tests/playwright/sanity/modules/nested-accordion/helper.js

Barcov committed:
Suggested Changes

Barcov committed:
Merge remote-tracking branch 'barcov/internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2

Barcov committed:
Suggested Changes

Barcov committed:
Merge branch 'main' into internal/ED-11088-NA_Internal_Tests_task_-_Create_E2E_tests2
